tools/outline_proxy_controller/outline_proxy_controller.cpp

```diff
@@ -276,6 +283,71 @@ void OutlineProxyController::detectBestInterfaceIndex() {
   }
 }
 
+bool OutlineProxyController::IsOutlineRoutingPolluted() noexcept {
+  if (routing_status_ == OutlineConnectionStatus::kReconfiguringRouting) {
+    return true;
+  }
+  if (routing_status_ != OutlineConnectionStatus::kRoutingThroughOutline) {
+    return false;
+  }
+
+  try {
+    auto result = executeIPRoute({});
+    if (!isSuccessful(result)) {
+      logger.warn("[routing polluted] failed to get routing table: " + result.first);
+      return true;
+    }
+
+    // A valid Outline'd `ip route` output must contains:
+    //   - One and only one default gateway through Outline `default via 10.0.85.2 dev outline-tun0`
+    //   - At least one non-Outline routing entry like `192.168.1.0/24 dev ens33 ...` (otherwise it
+    //     means the NIC might be turned off, and we need to enter "reconnecting" state)
+    bool has_outline_default_entry = false;
+    bool has_non_outline_device = false;
+
+    std::istringstream routing_table{result.first};
+    for (std::string routing_entry; std::getline(routing_table, routing_entry);) {
+      std::smatch entry_match;
+      if (std::regex_match(routing_entry, entry_match, kDefaultRoutingEntryPattern)) {
+        if (entry_match[1].str() == tunInterfaceRouterIp && entry_match[2].str() == tunInterfaceName) {
+          has_outline_default_entry = true;
+        } else {
+          logger.info("[routing polluted] extra non-Outline default gateway: " + routing_entry);
+          return true;
+        }
+      } else if (std::regex_match(routing_entry, entry_match, kRoutingEntryPattern)) {
+        if (entry_match[1].str() != tunInterfaceRouterIp && entry_match[2].str() != tunInterfaceName) {
+          has_non_outline_device = true;
+        }
+      }
+    }
+
+    if (!has_outline_default_entry || !has_non_outline_device) {
+      logger.info(std::string{"[routing polluted]"}
+        + (!has_outline_default_entry ? " no Outline default gateway;" : "")
+        + (!has_non_outline_device ? " no outgoing network interface;" : ""));
+      return true;
+    }
+    return false;
+  } catch (const std::exception &err) {
+    logger.warn("[routing polluted] unexpected error: " + std::string{err.what()});
+    return true;
+  }
+}
+
+bool OutlineProxyController::TryReconfigureRouting() noexcept {
```

Comment:
Remove `Try` from the name. It's not adding anything.
 
```suggestion
bool OutlineProxyController::ReconfigureRouting() noexcept {
```